### PR TITLE
refactor: Remove Item from merger's Node trait

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR removes the associated type `Item` from the `Node` trait of the merger mod. It is helpful to support `?Size` item of a node.

Now the `DataBatch` is a slice of the referencing `RecordBatch`.
- It doesn't own the `RecordBatch`.
- It should only access the range of the `RecordBatch`.
- `DataBatchRange` maintains the range of a `DataBatch`.

This avoids slicing the `RecordBatch` via `RecordBatch::slice()`.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)
- #2804

### CODE REVIEW

**Review by fengjiachun — COMMENTED** [review]
LGTM

**Review by fengjiachun — APPROVED** [review]

File: src/mito2/src/memtable/merge_tree/merger.rs [hunk 1]
```diff
@@ -17,55 +17,46 @@ use std::collections::BinaryHeap;
...
 pub struct Merger<T: Node> {
+    /// Heap to find node to read.
+    ///
+    /// Nodes in the heap are always valid.
     heap: BinaryHeap<T>,
-    current_item: Option<T::Item>,
+    /// Current node to read.
+    ///
+    /// The node is always empty if it is not None.
+    current_node: Option<T>,
```
Comment (v0y4g3r) [inline src/mito2/src/memtable/merge_tree/merger.rs:54]:
```suggestion
    /// The node is always **not** empty if it is not None.
```

File: src/mito2/src/memtable/merge_tree/merger.rs [hunk 2]
```diff
@@ -75,232 +66,173 @@ where
...
+            self.current_node = Some(node);
+        } else {
+            // Puts the node into the heap.
+            self.heap.push(node);
+        }
 
-impl PartialOrd<Self> for DataBatchKey {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        Some(self.cmp(other))
+        Ok(())
```
Comment (v0y4g3r) [inline src/mito2/src/memtable/merge_tree/merger.rs:164]: We can simplify this function to following, as the fast-path when heap is empty in which case `self.current_node` is set to current node happens only once per merger, but it brings more complexity.
```suggestion
        let Some(mut node) = self.current_node.take() else {
            return Ok(());
        };

        // Advances current node.
        node.advance(self.current_rows)?;
        self.current_rows = 0;
        if node.is_valid() {
            // Puts the node into the heap.
            self.heap.push(node);
        }
        Ok(())
```

Comment (v0y4g3r) [inline src/mito2/src/memtable/merge_tree/merger.rs:164]:
```suggestion
    fn maybe_advance_current_node(&mut self) -> Result<()> {
        let Some(mut node) = self.current_node.take() else {
            return Ok(());
        };

        // Advances current node.
        node.advance(self.current_rows)?;
        self.current_rows = 0;
        if node.is_valid() {
            // Puts the node into the heap.
            self.heap.push(node);
        }
        Ok(())
    }
```

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**